### PR TITLE
Support different types of tensor in the same sample

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

Support different types of tensor in the same sample

## How was this patch tested?

Added unit tests

### CODE REVIEW

**Comment by qiuxin2012** [comment]
Can you merge TensorSample and TableSample into one? I think they can be the same type of sample.

**Comment by yangw1234** [comment]
pr validation pass

**Comment by yangw1234** [comment]
@qiuxin2012 ready to be reviewed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 1]
```diff
@@ -396,34 +397,71 @@ object Sample {
...
 }
 
 /**
- * A kind of Sample who hold both DenseTensor and SparseTensor as features.
+ * A kind of Sample who holds different types of tensors as features.
  * @param features feature tensors
  * @param labels label tensors
  * @tparam T numeric type
  */
 private[bigdl] class TensorSample[T: ClassTag](
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: should we change the name to TableSample?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 2]
```diff
@@ -396,34 +397,71 @@ object Sample {
       ArraySample(featureTensors)
     }
   }
+
+  def apply[T: ClassTag](
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: The API looks not right. What's the meaning of T here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala [hunk 3]
```diff
@@ -396,34 +397,71 @@ object Sample {
...
+      featureTensors: Table)(implicit ev: TensorNumeric[T]) : Sample[T] = {
+    new TensorSample(featureTensors, T())
+  }
+
+  def apply[T: ClassTag](
+      featureTensors: Table,
+      labelTensors: Table)(implicit ev: TensorNumeric[T]) : Sample[T] = {
+    new TensorSample[T](featureTensors, labelTensors)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: save question with the above

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/MiniBatch.scala [hunk 4]
```diff
@@ -256,6 +256,10 @@ object MiniBatch {
     new ArrayTensorMiniBatch[T](input, target)
   }
 
+  def apply[T: ClassTag](input: Table, target: Table): MiniBatch[T] = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/MiniBatch.scala]: what's the meaning of T here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala (hunk 2 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: The base class Sample needs a type parameter T.

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Sample.scala]: I think we should change that. Or it's not a reasonable API.

**Pull request closed without merging**